File: Day9/solution.js
const { Console } = require('console');
const fs = require('fs');
const data = fs.readFileSync('./input.txt', 'utf-8').split('\r\n').map(Number);

// Part 1
function part1() {
    let invalidNum = false;
    let i = 0;

    while (invalidNum === false && i < data.length) {
        let arr = data.slice(i, i + 26);
        let checkNum = arr[arr.length - 1];
        let foundSum = false;

        for (let j = 0; j < arr.length - 2; j++) {
            let difference = Math.abs(arr[j] - checkNum);
            if (arr.includes(difference)) {
                foundSum = true;
                break;
            }
        }

        if (foundSum === false) {
            return checkNum;
        }

        i++;
    }
}

console.log('Part 1: ', part1());

// Part 2
function part2() {
    let invalidNum = part1();
    let sum = 0;
    let i = 0; 
    let j = 0;

    while (sum !== invalidNum) {
        while (sum < invalidNum) {
            sum += data[i]
            i++
        }
        
        while (sum > invalidNum) {
            sum -= data[j]
            j++
        }
    }

    let minmaxArray = data.slice(j,i).sort((a,b) => a - b);

    let min = Math.min(...minmaxArray)
    let max = Math.max(...minmaxArray)

    return min + max

}

console.log('Part 2: ', part2())

File: Day9/solution.py
data = [int(x) for x in open('./input.txt')]

# Part 1


def part1():
    invalid_number = False
    idx = 0

    while not invalid_number and idx < len(data):
        check_list = data[idx:idx+26]
        check_number = check_list[-1]
        found_sum = False

        for i in range(len(check_list)-1):
            diff = check_number - check_list[i]
            if diff in check_list:
                found_sum = True
                break

        if not found_sum:
            return check_number

        idx += 1


print(f'Part 1: {part1()}')


# Part 2
def part2():
    invalid_number = part1()
    sum = i = j = 0

    while sum != invalid_number:
        while sum < invalid_number:
            sum += data[i]
            i += 1
        while sum > invalid_number:
            sum -= data[j]
            j += 1

    sum_array = data[j:i]

    lowest = min(sum_array)
    highest = max(sum_array)

    return lowest + highest


print(f'Part 2: {part2()}')
